crates/uv-resolver/src/pubgrub/package.rs

```diff
@@ -60,6 +59,9 @@ pub enum PubGrubPackage {
         /// version before the registry version. So we could just error if we visit a URL variant
         /// _after_ a registry variant.
         Option<VerbatimUrl>,
+        /// Names of the file sources of this package (e.g. requirements.in), or empty for dependencies
+        /// only from other libraries
+        Vec<SourceName>,
```

Comment:
Could we instead omit this here, and re-connect the requirements to their paths when we build the `ResolutionGraph`?

Reply:
I think that could also enable correct annotation for (e.g.) `-c requirements.txt`, since then you could identify requirements that came from constraints.

Reply:
I started writing "here's why this doesn't work" and then found out in the course of writing it, that yes, it can be done! A tool I really want but can't seem to find with a bit of googling is some sort of code visualisation thing for Rust that'd let me visually see all of this v.s. the "lets cram it all into my head" approach I've been using, which clearly isn't the best option.

Here's my notes:

The only code that actually notices -r/-c requirements is `parse_entry` in the `requirements-txt` crate, which then only gets used by `parse`/`parse_inner` in `RequirementsTxt`. This results in path data in `RequirementEntry` and source values in the `Requirement`'s that are the `constraints` in the `RequirementsTxt` struct.

That in turn gets called by `RequirementsSpecification` `from_source`/`from_sources`, which similarly builds `requirements`/`constraints` with path/source data.

`pip_compile` in the `uv` crate then calls all that, feeds it into the pubgrub bits for resolution, and then feeds the results into `DisplayResolutionGraph` for display.

So, yeah. I could take the data from `RequirementsSpecification`, fish out a package -> sources hashmap and skip adding it to Pubgrub entirely... let me have a look at that.

Reply:
Fixed! Also resolved some test failures :)